### ISSUE TITLE
Add a rule to disallow "for … in" on array types

### ISSUE THREAD

**Issue by danvk** [palantir/tslint#1380]
This is almost certainly an error:

```typescript
function foo(values: number[]) {
  for (const value in values) {
    ...
  }
}
```

It should be "for … of". If the intention is to iterate over the array indices, that's better done with `values.forEach` or a traditional `for (let i = 0; i < values.length; i++)` loop. Both of these alternatives give you numeric indices, rather than the strings that "for ... in" does.

Some people advocate disallowing for-in entirely, but I still find it useful for objects.

In theory tslint should know whether the object being iterated over is really an array and it could warn specifically on this mistake.

**Comment by adidahiya** [comment]
This requires type checker information, which should be coming soon (#1363)!

**Comment by danvk** [comment]
I've sent out a PR to implement this.
